url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/MvdTool.java

```diff
@@ -0,0 +1,975 @@
+		}
+	}
+	/**
+	 * List the versions of an MVD
+	 * @throws MVDToolException
+	 */
+	private static void doListVersions() throws MVDToolException
+	{
+		try
+		{
+			MVD mvd = loadMVD();
+			String versionTable = mvd.getVersionTable();
+			out.write( versionTable.getBytes(mvd.getEncoding()) );
+		}
+		catch ( Exception e )
+		{
+			throw new MVDToolException( e );
+		}
+	}
+	/**
+	 * Add the specified version to the MVD. Don't replace an existing 
+	 * version, so we reset the supplied version parameter to the number 
+	 * of versions+1
+	 */
+	private static void doAddVersion() throws MVDToolException
+	{
+		try
+		{
+			File t = new File( textFile );
+			if ( t.exists() )
+			{
+				File m = new File( mvdFile );
+				MVD mvd;
+				if ( mvdFile != null )
+				{
+					mvd = MVDFile.internalise( m, Utilities.loadDBProperties(dbConn) );
+                    mvd.setDirectAlign( directAlignOnly );
+					mvd.newVersion(shortName, longName, groupName, 
+						(short)backup, partial );
+				}
+				else
+				{
+					mvd = new MVD( description );
+                    mvd.setDirectAlign( directAlignOnly );
+					mvd.newVersion( shortName, longName, groupName, 
+						(short)backup, partial );
+				}
+				FileInputStream fis = new FileInputStream( t );
+				byte[] data = new byte[(int)t.length()];
+				fis.read( data );
+				char[] chars = Utilities.bytesToChars(data,mvd.getEncoding());
+                version = (short)(mvd.numVersions());
+				mvd.update( version, chars, mergeSharedVersions );
+				if ( mvd.getDescription() != null )
+				{
+					MVDFile.externalise( mvd, m, folderId, 
+						Utilities.loadDBProperties(dbConn) );
+				}
+				MVDError.log("Unique percentage="+mvd.getUniquePercentage(version));
+				MvdTool.out.println(mvd.getUniquePercentage(version));
+			}
+			else
+				throw new FileNotFoundException(
+					"Couldn't find file "+t.getAbsolutePath() );
+		}
+		catch ( Exception e )
+		{
+			e.printStackTrace();
+			throw new MVDToolException( e );
+		}
+	}
+	/**
+	 * Write out all the versions as separate files
+	 */
+	private static void doArchive() throws MVDException
+	{
+		try
+		{
+			MVD mvd = loadMVD();
+			File m = new File( mvdFile );
+			File archiveDir;
+			XMLGuideFile guideFile = new XMLGuideFile( mvd );
+			// create default archive name
+			if ( archiveName == null )
+			{
+				archiveName = m.getName();
+				int index = archiveName.lastIndexOf(".");
+				archiveName = archiveName.substring( 0, index );
+				archiveDir = new File( m.getParentFile(), archiveName );
+			}
+			else
+				archiveDir = new File( archiveName );
+			if ( !archiveDir.exists() )
+				archiveDir.mkdir();
+			int nVersions = mvd.numVersions();
+			for ( short i=1;i<=nVersions;i++ )
+			{
+				char[] data = mvd.getVersion( i );
+				// use the short name as the file name
+				String vName = mvd.getVersionShortName( i );
+				File versionFile = new File( archiveDir, vName );
+                FileOutputStream fos = new FileOutputStream( versionFile );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at lines 622, 623. The resource is closed at line 624. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).